### PR TITLE
feat: make the pytest testing framework functional, migrate all v2 tests

### PR DESCRIPTION
**Pull request by fzipi**
## Summary

Adds a pytest-based test framework under `tests/`, alongside the existing Perl-based `test/` and `tests/run-regression-tests.pl` infrastructure. It covers the same test data (`tests/op/*.t`, `tests/tfn/*.t`, `tests/regression/*/*.t`) rather than introducing new test cases, with two exceptions noted below - those `.t` files (where kept, see below) remain the source of truth.

- **Unit tests** (`tests/test_operators/`, `tests/test_transformations/`): run each `@operator`/`t:transformation` case through the `msc_test` C binary. No Apache needed. 4228 cases. These were migrated from `tests/op/*.t`/`tests/tfn/*.t` via a one-time Perl-assisted conversion, then the `.t` files and conversion tooling were removed - that data is flat literal strings with no ongoing need for a Perl round-trip, so the generated `.py` files are now hand-maintained directly (add/edit a parametrize tuple like any other pytest file). Two of these files, `test_beginswith.py` and `test_base64decode.py`, carry hand-added edge-case coverage (Unicode input, long strings, binary data) beyond what their source `.t` file had - pre-existing from this framework's initial scaffolding, not new in this PR, but worth calling out against the "same test data" framing above. `tests/run-unit-tests.pl` (the Perl runner for this same data) is removed along with its `Makefile.am`/`configure.ac` wiring, since it has no data left to run against; `make check`/`make test` still succeed, just with nothing left to run in `tests/`.
- **Regression tests** (`tests/test_regression/`): run against a live Apache + `mod_security2`, reusing the autoconf-generated `tests/regression/server_root/conf/httpd.conf`. Unlike unit tests, `tests/regression/*/*.t` stays as the permanent source of truth - `qr//` regex flags, `HTTP::Request` construction, and occasional `conf => sub {...}` coderefs are Perl semantics worth keeping Perl around to parse rather than hand-maintaining in a flattened form. `dump_regression_fixtures.pl` evaluates each `.t` file the same way `run-regression-tests.pl` does (same `%ENV`, same eval) and serializes the result to `tests/regression/fixtures/*.json`, with this-machine absolute paths replaced by portable `##PLACEHOLDER##` tokens that `regression_fixtures.py` resolves back to local paths at load time; `test_regression/test_fixtures.py` discovers and runs every entry in those fixtures, one pytest case each - no per-`.t`-file Python to hand-maintain. `tests/regenerate_regression_fixtures.sh` regenerates the fixtures after editing a `.t` file; CI fails if they're stale. 246 cases: 241 pass, 5 skipped with a stated reason (1 relies on a Perl coderef this migration doesn't execute; 4 check exact HTTP request-header bytes that depend on the client library's fingerprint - `requests`/`urllib3` vs. the original `LWP::UserAgent`).
- `.github/workflows/ci.yml`: `test-linux` and `test-regression-linux` now build/run these pytest suites as part of their existing steps, in place of the `make test`/`make test-regression` invocations that drove the Perl runners.
- `tools/legacy/`: two Perl scripts (`gen_rx-pm.pl.in`, `csv_rx-pm.pl.in`) moved here from `tests/` - they were only ever auto-generated by `configure`, never invoked by any Makefile target or CI step.
- `sonar-project.properties`: scopes SonarCloud's hardcoded-IP-address rule out of `tests/test_operators/`/`tests/test_transformations/`, since that's literal test data for the `@ipMatch`/`@geoLookup` operators, not a real code smell.
- `tests/README.md` documents the structure above and how to run/regenerate each suite.

## Test plan

- [x] `pytest tests/test_operators tests/test_transformations` - 4228 passed
- [x] `pytest tests/test_regression` - 241 passed, 5 skipped (reasons documented in `tests/README.md`)
- [x] Combined `pytest` run from `tests/` - 4469 passed, 5 skipped
- [x] `tests/regenerate_regression_fixtures.sh` produces byte-identical output to what's checked in (verifies the CI staleness check)
- [x] Verified no checked-in fixture contains a this-machine absolute path (`grep -rl` across `tests/regression/fixtures/` for the local checkout path)
- [x] `autogen.sh`/`configure` still succeed after removing `run-unit-tests.pl(.in)` and moving the two legacy Perl scripts
- [x] `make check` in `tests/` still succeeds (builds `msc_test`, runs no tests there now)
- [x] `actionlint`/`zizmor` on the updated `ci.yml` show no new findings beyond the file's pre-existing style

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by airween** [comment]
I like the idea of ​​extending/evolving the testing framework, and besides my preference for Python over Perl :smiley:, now I think that instead of replacing the entire existing Perl-based regression testing framework, I would add this new one beside the existing one; which could run in parallel.

The reason is very simple: there are a few Sonar issues, probably we will fix them over time, and there is one regression test which can't be completed. I've restarted it, but the situation is the same.

In short: I would give time to improve the new framework.

What do you think?

**Comment by fzipi** [comment]
Of course, this is my first PR 😉 

The idea is to evolve it a bit and replace. It doesn't make sense to have them both in parallel.

Will do more work.

File: tests/pytest.ini [hunk 1]
```diff
@@ -0,0 +1,21 @@
+[tool:pytest]
```
[the remark this answers is not in the document]
Comment (fzipi) [reply tests/pytest.ini]: Fixed in 3846fc9d - changed the section header to `[pytest]`. Confirmed empirically that `--strict-markers` and `addopts` were being silently ignored before this fix.

File: tests/regression/fixtures/rule/20-exceptions.json [hunk 2]
```diff
@@ -0,0 +1 @@
..._regex__":1,"flags":"","pattern":"500002"},1],"debug":[{"__regex__":1,"flags":"","pattern":"id:500002,pass,nolog"},1]},"match_response":{"status":{"__regex__":1,"flags":"","pattern":"^200$"}},"request":{"__type__":"http_request","content":"","headers":[],"method":"GET","uri":"http://localhost:8088/test.txt"},"type":"rule"},{"comment":"SecRuleUpdateActionById (chain)","conf":"\n\t\tSecRuleEngine On\n\t\tSecDebugLog /Users/fzipitria/Workspace/OWASP/ModSecurity/tests/regression/server_root/logs/modsec_debug.log\n\t\tSecDebugLogLevel 9\n\t\tSecRule REQUEST_URI \"test\" \"phase:1,deny,status:500,msg:'testing rule',chain,id:500003\"\n        SecRule ARGS \"bar\"\n\t\tSecRuleUpdateActionById 500003 \"pass,nolog\"\n\t","match_log":{"-audit":[{"__regex__":1,"flags":"","pattern":"."},1],"-debug":[{"__regex__":1,"flags":"","pattern":"Access denied"},1],"-error":[{"__regex__":1,"flags":"","pattern":"500003"},1],"debug":[{"__regex__":1,"flags":"","pattern":"id:500003,pass,nolog"},1]},"match_response":{"status":{"__regex__":1,"flags":"","pattern":"^200$"}},"request":{"__type__":"http_request","content":"","headers":[],"method":"GET","uri":"http://localhost:8088/test.txt?foo=bar"},"type":"rule"}]
```
[the remark this answers is not in the document]
Comment (fzipi) [reply tests/regression/fixtures/rule/20-exceptions.json]: Fixed in 3846fc9d - `dump_regression_fixtures.pl` now swaps this-machine absolute paths for portable `##KEY##` placeholders before writing the JSON, and `regression_fixtures.py` resolves them back to local paths at load time. Verified no absolute local paths remain in any checked-in fixture.

File: tests/regression/fixtures/config/10-misc-directives.json [hunk 3]
```diff
@@ -0,0 +1 @@
...0tMTk4MTMxODE3NzE4MzA3NjU2NDM5OTYxODcyMDYtLQ==","headers":[["Content-Type","multipart/form-data; boundary=---------------------------19813181771830765643996187206"]],"method":"POST","uri":"http://localhost:8088/test.txt"},"type":"config","unsupported":"test"},{"comment":"SecWebAppId","conf":"\n\t\tSecRuleEngine On\n\t\tSecRequestBodyAccess On\n\t\tSecDebugLog /Users/fzipitria/Workspace/OWASP/ModSecurity/tests/regression/server_root/logs/modsec_debug.log\n\t\tSecDebugLogLevel 4\n\t\tSecAuditLog \"/Users/fzipitria/Workspace/OWASP/ModSecurity/tests/regression/server_root/logs/modsec_audit.log\"\n\t\tSecAuditEngine RelevantOnly\n\t\tSecWebAppId \"app-1\"\n\t\tSecAction \"pass,log,auditlog,id:1\"\n\t","match_log":{"audit":[{"__regex__":1,"flags":"m","pattern":"^WebApp-Info: \"app-1\""},1],"debug":[{"__regex__":1,"flags":"","pattern":"Warning\\. Unconditional match in SecAction\\."},1],"error":[{"__regex__":1,"flags":"","pattern":"Warning\\. Unconditional match in SecAction\\."},1]},"match_response":{"status":{"__regex__":1,"flags":"","pattern":"^200$"}},"request":{"__type__":"http_request","content":"","headers":[],"method":"GET","uri":"http://localhost:8088/test.txt"},"type":"config"}]
```
[the remark this answers is not in the document]
Comment (fzipi) [reply tests/regression/fixtures/config/10-misc-directives.json]: Fixed in 3846fc9d via the same placeholder mechanism (see reply on the 20-exceptions.json thread).

File: tests/dump_regression_fixtures.pl [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    REGRESSION_DIR => $REG_DIR,
+    DIST_ROOT => File::Spec->rel2abs(dirname("$SCRIPT_DIR/../..")),
+    AUDIT_LOG => "$FILES_DIR/modsec_audit.log",
+    DEBUG_LOG => "$FILES_DIR/modsec_debug.log",
+    ERROR_LOG => "$FILES_DIR/error.log",
+    HTTPD_CONF => "$CONF_DIR/httpd.conf",
+    HTDOCS => "$SROOT_DIR/htdocs",
+    USER_AGENT => "ModSecurity Regression Tests/1.2.3",
+    RUNASUSER => $ENV{USER} || $ENV{LOGNAME} || $ENV{USERNAME} || 'unknown',
+);
```
[the remark this answers is not in the document]
Comment (fzipi) [reply tests/dump_regression_fixtures.pl:83]: Fixed in 3846fc9d - the dumper still evaluates `.t` files with real local paths in `%ENV` (so `conf => sub {...}` coderefs that do real file I/O keep working), but now substitutes this-machine paths for portable `##KEY##` placeholders before emitting the final JSON, longest value first so a shorter path that's a prefix of a longer one doesn't break the match.

File: tests/test_transformations/test_base64decode.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                "name": "base64Decode",
+                "input": case["input"],
+                "output": case["output"],
+                "ret": case["ret"]
+            }
+            
+            result = unit_test.run_unit_test(test_config)
+            # Note: Some invalid cases might behave differently than expected
+            # This is testing the actual behavior rather than ideal behavior
+            assert result.success or case["ret"] == 0, f"Invalid input test failed: {case['desc']}"
```
[the remark this answers is not in the document]
Comment (fzipi) [reply tests/test_transformations/test_base64decode.py]: Fixed in 3846fc9d - the always-true `or case["ret"] == 0` fallback masked a real bug. Verified against `msc_test` directly: `base64Decode` lenient-decodes valid-looking characters rather than failing outright, producing specific non-empty bytes for this input, not "" with ret=0. Corrected the expected data and removed the fallback.

File: tests/test_operators/test_beginswith.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    
+    # Special characters
+    ("test@example.com", "test@example.com/path", 1, "Email prefix match"),
+    ("http://", "http://example.com", 1, "URL prefix match"),
+    ("hello world", "hello world test", 1, "Space in param"),
+    
+    # Unicode tests
+    ("café", "café au lait", 1, "Unicode prefix match"),
+    ("测试", "测试数据", 1, "Chinese characters"),
+])
```
[the remark this answers is not in the document]
Comment (fzipi) [reply tests/test_operators/test_beginswith.py:28]: Confirmed this predates the PR (part of the framework's initial scaffolding) rather than being introduced by this PR, but the PR description's "same test data, no new cases" framing was inaccurate either way. Fixed in 3846fc9d by calling out both `test_beginswith.py` and `test_base64decode.py` explicitly in the description as carrying hand-added edge-case coverage beyond their source `.t` file. Keeping the extra coverage since it's real, verified-against-`msc_test` behavior, not fabricated.

File: tests/apache_server.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+        self._per_test_conf: Optional[Path] = None
+
+    def _run_control(self, extra_conf: Optional[str], action: str) -> subprocess.CompletedProcess:
+        cmd = [
+            self.httpd_path,
+            "-d", self.apache_server_root,
+            "-f", str(self.base_conf),
+            "-c", f"Listen {self.server_name}:{self.port}",
+        ]
```
[the remark this answers is not in the document]
Comment (fzipi) [reply tests/apache_server.py:82]: Fixed in 3846fc9d - `-d` is now only passed to httpd when `_resolve_server_root()` actually found a non-empty ServerRoot.

**Comment by fzipi** [comment]
The `E` maintainability is a false positive here. Tests have IPs in fixtures, and SonarQube wants you to don't hardcode IPs e.g.: `IP addresses should not be hardcoded [python:S1313](https://sonarcloud.io/organizations/owasp-modsecurity/rules?open=python%3AS1313&rule_key=python%3AS1313)`

**Comment by fzipi** [comment]
Did tried to add this file: https://github.com/fzipi/ModSecurity/blob/6c8c656020b76308b36d6f98f76b62147f34fcaa/.sonarcloud.properties, but looks like sonarcloud is not processing it. Maybe this can be enabled at the Org level @airween ?

**Comment by airween** [comment]
> Did tried to add this file: https://github.com/fzipi/ModSecurity/blob/6c8c656020b76308b36d6f98f76b62147f34fcaa/.sonarcloud.properties, but looks like sonarcloud is not processing it. Maybe this can be enabled at the Org level @airween ?

I think it's already installed at org level:
https://github.com/organizations/owasp-modsecurity/settings/installations/46554971
